File: plugins/bubblellaneous/registry/miscellaneous.py
from plugins.bubblellaneous.internal import Block, BlockData, BlockMaterials
from plugins.bubblellaneous.internal.unit.block import BlockType
from plugins.bubblellaneous.internal.unit.item import Item, ItemData
from plugins.bubblellaneous.internal.unit.variant import BlockVariant, ItemVariant


class WindowBoards(BlockVariant):
    """
    :yellow [☶ Description]
    Empty.

    :light_purple [⇄ Connectivity]
    Can connect to other window boards.
    """

    material = BlockMaterials.SOLID

    base = BlockData.Base.VOID
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.WALL_NORMAL
    recipe = [
        BlockData.RecipeEntry("block", "[base]", 4),
    ]
    tags = [BlockData.Uses.BLOCKSTATES]
    blockstates = BlockData.BlockStates(
        "@self",
        BlockData.State("default"),
        BlockData.State("left", "~~01~~"),
        BlockData.State("right", "~~10~~"),
        BlockData.State("middle", "~~11~~"),
    )


class Hatch(Block):
    """
    :yellow [☶ Description]
    Click to open/close.
    """

    base = BlockData.Base.VOID
    sound = BlockData.Sound.INDUSTRIAL
    facing = BlockData.Facing.NONE
    recipe = [
        BlockData.RecipeEntry("item", "iron_ingot", 4),
    ]
    tags = [
        BlockData.Uses.PLACE,
        BlockData.Uses.NO_BASE,
        BlockData.Uses.CUSTOM_BASE,
        BlockData.Uses.TIMER,
    ]
    blockstates = BlockData.BlockStates(
        "<manual>",
        BlockData.State("default"),
        BlockData.State("rim"),
        BlockData.State("latch"),
    )


class TrafficBarrier(Block):
    """
    :yellow [☶ Description]
    Empty.
    """

    children = ["block/mossy_traffic_barrier"]
    base = BlockData.Base.UPPER
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.PLAYER_PRECISE
    recipe = [
        BlockData.RecipeEntry("block", "oak_planks", 6),
    ]
    tags = []


class MossyTrafficBarrier(Block):
    """
    :yellow [☶ Description]
    Empty.
    """

    is_unlisted = True
    base = BlockData.Base.UPPER
    sound = BlockData.Sound.WOOD
    facing = BlockData.Facing.PLAYER_PRECISE
    recipe = [
        BlockData.RecipeEntry("block", "oak_planks", 6),
        BlockData.RecipeEntry("block", "moss_carpet", 1),
    ]
    tags = []


class Pole(Block):
    """
    :yellow [☶ Description]
    Empty.
    """

    base = BlockData.Base.CHAIN
    sound = BlockData.Sound.INDUSTRIAL
    facing = BlockData.Facing.NONE
    recipe = [
        BlockData.RecipeEntry("block", "chain", 1),
    ]
    tags = []


class RoadSign(Block):
    """
    :yellow [☶ Description]
    Empty.
    """

    base = BlockData.Base.CHAIN
    sound = BlockData.Sound.INDUSTRIAL
    facing = BlockData.Facing.PLAYER
    recipe = [
        BlockData.RecipeEntry("block", "chain", 1),
        BlockData.RecipeEntry("block", "white_banner", 1),
    ]
    tags = []


class NoEntranceRoadSign(RoadSign):
    children = [
        "block/park_road_sign",
        "block/arrow_down_road_sign",
        "block/arrow_up_road_sign",
        "block/arrow_left_road_sign",
        "block/arrow_right_road_sign",
    ]


class ParkRoadSign(RoadSign):
    is_unlisted = True


class ArrowDownRoadSign(RoadSign):
    is_unlisted = True


class ArrowUpRoadSign(RoadSign):
    is_unlisted = True


class ArrowLeftRoadSign(RoadSign):
    is_unlisted = True


class ArrowRightRoadSign(RoadSign):
    is_unlisted = True


class CallBell(Block):
    """
    :yellow [☶ Description]
    Click to play bell sound.
    """

    base = BlockData.Base.VOID
    sound = BlockData.Sound.INDUSTRIAL
    facing = BlockData.Facing.NONE
    recipe = [
        BlockData.RecipeEntry("item", "iron_ingot", 1),
    ]
    tags = [
        BlockData.Uses.PLACE,
        BlockData.Uses.NO_BASE,
        BlockData.Uses.CUSTOM_BASE,
    ]


class Box(Block):
    """
    :yellow [☶ Description]
    A regular container (27 slots).
    """

    base = BlockData.Base.CONTAINER
    sound = BlockData.Sound.WOOL
    facing = BlockData.Facing.PLAYER
    recipe = [
        BlockData.RecipeEntry("item", "leather", 8),
    ]
    tags = [BlockData.Uses.BRIGHTNESS_FIX]


class VendingMachine(Block):
    """
    :yellow [☶ Description]
    Empty.

    :light_purple [⇄ Connectivity]
    Can connect to other vending machines.
    """

    base = BlockData.Base.SOLID
    sound = BlockData.Sound.INDUSTRIAL
    facing = BlockData.Facing.PLAYER
    recipe = [
        BlockData.RecipeEntry("block", "glass", 1),
        BlockData.RecipeEntry("item", "iron_ingot", 2),
    ]
    tags = [
        BlockData.Uses.BRIGHTNESS_FIX,
        BlockData.Uses.BLOCKSTATES,
    ]
    blockstates = BlockData.BlockStates(
        "@self",
        BlockData.State("default"),
        BlockData.State("single_left", "~~1000"),
        BlockData.State("single_middle", "~~1100"),
        BlockData.State("single_right", "~~0100"),
        BlockData.State("top_single", "~~0001"),
        BlockData.State("top_left", "~~1001"),
        BlockData.State("top_middle", "~~1101"),
        BlockData.State("top_right", "~~0101"),
        BlockData.State("middle_single", "~~0011"),
        BlockData.State("middle_left", "~~1011"),
        BlockData.State("middle_middle", "~~1111"),
        BlockData.State("middle_right", "~~0111"),
        BlockData.State("bottom_single", "~~0010"),
        BlockData.State("bottom_left", "~~1010"),
        BlockData.State("bottom_middle", "~~1110"),
        BlockData.State("bottom_right", "~~0110"),
    )


class Megaphone(Item):
    """
    :yellow [☶ Description]
    Empty.
    """

    base = ItemData.Base.INTERACTIVE
    recipe = [
        BlockData.RecipeEntry("item", "iron_ingot", 1),
    ]
    params = lambda *_: {}


class Cup(BlockVariant):
    """
    :yellow [☶ Description]
    Empty.
    """

    material = BlockMaterials.DYE

    base = BlockData.Base.VOID
    sound = BlockData.Sound.WOOL
    facing = BlockData.Facing.PLAYER_PRECISE
    recipe = [
        BlockData.RecipeEntry("item", "[color]", 3),
    ]
    tags = []


class Clipboard(Block):
    """
    :yellow [☶ Description]
    Empty.
    """

    base = BlockData.Base.VOID
    sound = BlockData.Sound.WOOL
    facing = BlockData.Facing.PLAYER_PRECISE
    recipe = [
        BlockData.RecipeEntry("item", "paper", 1),
    ]
    tags = []


class Plate(Block):
    """
    :yellow [☶ Description]
    Can store a single item stack.
    """

    block_type = BlockType.shelf(amount=1, offset=-0.45)

    base = BlockData.Base.VOID
    sound = BlockData.Sound.WOOL
    facing = BlockData.Facing.PLAYER_PRECISE
    recipe = [
        BlockData.RecipeEntry("block", "quartz_block", 1),
    ]
    tags = [BlockData.Uses.PLACE]


class Speakers(Block):
    """
    :yellow [☶ Description]
    Empty.
    """

    base = BlockData.Base.SOLID
    sound = BlockData.Sound.INDUSTRIAL
    facing = BlockData.Facing.PLAYER
    recipe = [
        BlockData.RecipeEntry("block", "black_wool", 1),
    ]
    tags = [BlockData.Uses.BRIGHTNESS_FIX]


class AlarmClock(Block):
    """
    :yellow [☶ Description]
    Empty.
    """

    base = BlockData.Base.VOID
    sound = BlockData.Sound.INDUSTRIAL
    facing = BlockData.Facing.PLAYER_PRECISE
    recipe = [
        BlockData.RecipeEntry("block", "bell"),
    ]
    tags = []


class Umbrella(Item):
    """
    :yellow [☶ Description]
    Hold Right Click to get slow falling.
    """

    base = ItemData.Base.INTERACTIVE
    recipe = [
        BlockData.RecipeEntry("item", "leather", 4),
    ]
    params = lambda *_: {}


class BearTrap(Block):
    """
    :yellow [☶ Description]
    Traps a player that steps on top of it for a set period of time.
    """

    base = BlockData.Base.TRIPWIRE
    sound = BlockData.Sound.INDUSTRIAL
    facing = BlockData.Facing.PLAYER
    recipe = [
        BlockData.RecipeEntry("item", "iron_ingot", 2),
    ]
    tags = [
        BlockData.Uses.TICK,
        BlockData.Uses.TIMER,
    ]
    blockstates = BlockData.BlockStates(
        "<manual>",
        BlockData.State("default"),
        BlockData.State("closed"),
    )


class PlayingCard(ItemVariant):
    """
    :yellow [☶ Description]
    Empty.
    """

    material = BlockMaterials.CARD_DECK

    base = ItemData.Base.NORMAL
    recipe = [
        BlockData.RecipeEntry("item", "paper", 1),
    ]
    params = lambda *_: {}

File: bubblellaneous/miscellaneous.go
package bubblellaneous

import (
	"github.com/bbfh-dev/bubblellaneous-pack/bubblellaneous/unit"
	"github.com/bbfh-dev/bubblellaneous-pack/lib/field"
	"github.com/bbfh-dev/bubblellaneous-pack/lib/nbt"
)

var Miscellaneous = unit.NewCategory().
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"window_boards",
			field.BASE_VOID,
			field.SOUND_WOOD,
			field.FACING_WALL_NORMAL,
			[]field.RecipeEntry{
				field.NewRecipeEntry("block", "[base]", 4),
			},
			field.USE_BLOCKSTATES,
		).WithBlockstates(
			"@self", field.NewBlockState("default"),
			field.NewBlockState("left", "~~01~~"),
			field.NewBlockState("right", "~~10~~"),
			field.NewBlockState("middle", "~~11~~"),
		)
	}, field.SOLID_MATERIAL).Units()...).
	Add(unit.NewBlock(
		"hatch",
		field.BASE_VOID,
		field.SOUND_INDUSTRIAL,
		field.FACING_NONE,
		[]field.RecipeEntry{
			field.NewRecipeEntry("item", "iron_ingot", 4),
		},
		field.USE_PLACE,
		field.USE_NO_BASE,
		field.USE_CUSTOM_BASE,
		field.USE_TIMER,
	).WithBlockstates(
		"<manual>", field.NewBlockState("default"),
		field.NewBlockState("rim"),
		field.NewBlockState("latch"),
	)).
	Add(unit.NewBlock(
		"traffic_barrier",
		field.BASE_UPPER,
		field.SOUND_WOOD,
		field.FACING_PLAYER_PRECISE,
		[]field.RecipeEntry{
			field.NewRecipeEntry("block", "oak_planks", 6),
		},
	)).
	Add(unit.NewBlock(
		"mossy_traffic_barrier",
		field.BASE_UPPER,
		field.SOUND_WOOD,
		field.FACING_PLAYER_PRECISE,
		[]field.RecipeEntry{
			field.NewRecipeEntry("block", "oak_planks", 6),
			field.NewRecipeEntry("block", "moss_carpet", 1),
		},
	)).
	Add(unit.NewBlock(
		"pole",
		field.BASE_CHAIN,
		field.SOUND_INDUSTRIAL,
		field.FACING_NONE,
		[]field.RecipeEntry{
			field.NewRecipeEntry("block", "chain", 1),
		},
	)).
	Add(unit.NewBlock(
		"box",
		field.BASE_CONTAINER,
		field.SOUND_WOOL,
		field.FACING_PLAYER,
		[]field.RecipeEntry{
			field.NewRecipeEntry("item", "leather", 8),
		},
		field.USE_BRIGHTNESS_FIX,
	)).
	Add(unit.NewBlock(
		"call_bell",
		field.BASE_VOID,
		field.SOUND_INDUSTRIAL,
		field.FACING_NONE,
		[]field.RecipeEntry{
			field.NewRecipeEntry("item", "iron_ingot", 1),
		},
		field.USE_PLACE,
		field.USE_NO_BASE,
		field.USE_CUSTOM_BASE,
	)).
	Add(unit.NewBlock(
		"vending_machine",
		field.BASE_SOLID,
		field.SOUND_INDUSTRIAL,
		field.FACING_PLAYER,
		[]field.RecipeEntry{
			field.NewRecipeEntry("block", "glass", 1),
			field.NewRecipeEntry("item", "iron_ingot", 2),
		},
		field.USE_BRIGHTNESS_FIX,
		field.USE_BLOCKSTATES,
	).WithBlockstates(
		"@self", field.NewBlockState("default"),
		field.NewBlockState("single_left", "~~1000"),
		field.NewBlockState("single_middle", "~~1100"),
		field.NewBlockState("single_right", "~~0100"),
		field.NewBlockState("top_single", "~~0001"),
		field.NewBlockState("top_left", "~~1001"),
		field.NewBlockState("top_middle", "~~1101"),
		field.NewBlockState("top_right", "~~0101"),
		field.NewBlockState("middle_single", "~~0011"),
		field.NewBlockState("middle_left", "~~1011"),
		field.NewBlockState("middle_middle", "~~1111"),
		field.NewBlockState("middle_right", "~~0111"),
		field.NewBlockState("bottom_single", "~~0010"),
		field.NewBlockState("bottom_left", "~~1010"),
		field.NewBlockState("bottom_middle", "~~1110"),
		field.NewBlockState("bottom_right", "~~0110"),
	)).
	Add(unit.NewItem(
		"megaphone",
		field.ITEM_INTERACTIVE,
		[]field.RecipeEntry{
			field.NewRecipeEntry("item", "iron_ingot", 1),
		},
		nbt.Tree(),
	)).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewBlock(
			"cup",
			field.BASE_VOID,
			field.SOUND_WOOL,
			field.FACING_PLAYER_PRECISE,
			[]field.RecipeEntry{
				field.NewRecipeEntry("item", "[color]", 3),
			},
		)
	}, field.DYE_MATERIAL).Units()...).
	Add(unit.NewBlock(
		"clipboard",
		field.BASE_VOID,
		field.SOUND_WOOL,
		field.FACING_PLAYER_PRECISE,
		[]field.RecipeEntry{
			field.NewRecipeEntry("item", "paper", 1),
		},
	)).
	Add(unit.NewBlock(
		"plate",
		field.BASE_VOID,
		field.SOUND_WOOL,
		field.FACING_PLAYER_PRECISE,
		[]field.RecipeEntry{
			field.NewRecipeEntry("block", "quartz_block", 1),
		},
		field.USE_PLACE,
	).WithBlockType(field.ShelfBlock(1, -0.45))).
	Add(unit.NewBlock(
		"speakers",
		field.BASE_SOLID,
		field.SOUND_INDUSTRIAL,
		field.FACING_PLAYER,
		[]field.RecipeEntry{
			field.NewRecipeEntry("block", "black_wool", 1),
		},
		field.USE_BRIGHTNESS_FIX,
	)).
	Add(unit.NewBlock(
		"alarm_clock",
		field.BASE_VOID,
		field.SOUND_INDUSTRIAL,
		field.FACING_PLAYER_PRECISE,
		[]field.RecipeEntry{
			field.NewRecipeEntry("block", "bell", 1),
		},
	)).
	Add(unit.NewItem(
		"umbrella",
		field.ITEM_INTERACTIVE,
		[]field.RecipeEntry{
			field.NewRecipeEntry("item", "leather", 4),
		},
		nbt.Tree(),
	)).
	Add(unit.NewBlock(
		"bear_trap",
		field.BASE_TRIPWIRE,
		field.SOUND_INDUSTRIAL,
		field.FACING_PLAYER,
		[]field.RecipeEntry{
			field.NewRecipeEntry("item", "iron_ingot", 2),
		},
		field.USE_TICK,
		field.USE_TIMER,
	).WithBlockstates(
		"<manual>", field.NewBlockState("default"),
		field.NewBlockState("closed"),
	)).
	Add(unit.NewBlock(
		"no_entrance_road_sign",
		field.BASE_CHAIN,
		field.SOUND_INDUSTRIAL,
		field.FACING_PLAYER,
		[]field.RecipeEntry{
			field.NewRecipeEntry("block", "chain", 1),
			field.NewRecipeEntry("block", "white_banner", 1),
		},
	).WithGroupId("road_sign")).
	Add(unit.NewBlock(
		"arrow_down_road_sign",
		field.BASE_CHAIN,
		field.SOUND_INDUSTRIAL,
		field.FACING_PLAYER,
		[]field.RecipeEntry{
			field.NewRecipeEntry("block", "chain", 1),
			field.NewRecipeEntry("block", "white_banner", 1),
		},
	).WithGroupId("road_sign")).
	Add(unit.NewBlock(
		"arrow_up_road_sign",
		field.BASE_CHAIN,
		field.SOUND_INDUSTRIAL,
		field.FACING_PLAYER,
		[]field.RecipeEntry{
			field.NewRecipeEntry("block", "chain", 1),
			field.NewRecipeEntry("block", "white_banner", 1),
		},
	).WithGroupId("road_sign")).
	Add(unit.NewBlock(
		"arrow_left_road_sign",
		field.BASE_CHAIN,
		field.SOUND_INDUSTRIAL,
		field.FACING_PLAYER,
		[]field.RecipeEntry{
			field.NewRecipeEntry("block", "chain", 1),
			field.NewRecipeEntry("block", "white_banner", 1),
		},
	).WithGroupId("road_sign")).
	Add(unit.NewBlock(
		"arrow_right_road_sign",
		field.BASE_CHAIN,
		field.SOUND_INDUSTRIAL,
		field.FACING_PLAYER,
		[]field.RecipeEntry{
			field.NewRecipeEntry("block", "chain", 1),
			field.NewRecipeEntry("block", "white_banner", 1),
		},
	).WithGroupId("road_sign")).
	Add(unit.NewVariant(func() unit.Unit {
		return unit.NewItem(
			"playing_card",
			field.ITEM_NORMAL,
			[]field.RecipeEntry{
				field.NewRecipeEntry("item", "paper", 1),
			},
			nbt.Tree(),
		)
	}, field.CARD_DECK_MATERIAL()).Units()...)
